### PR TITLE
Help: Track eroneous sent/received chat messages

### PR DESCRIPTION
**Pull request by omarjackman**
In general chat messages shouldn't be sent/received while on the contact form without the chatbox being inlined. This pull request adds event handlers that track the occurrence of these messages and adds information about the state of olark.

Testing this is tricky since the issue that this pull request aims to track is not generally reproducible.

### CODE REVIEW

**Comment by omarjackman** [comment]
cc @dllh 
thoughts, comments or criticisms?

File: client/me/help/help-contact/index.jsx [hunk 1]
```diff
@@ -218,6 +222,33 @@ module.exports = React.createClass( {
 		widgetInput.onkeydown( { keyCode: KEY_ENTER } );
 	},
 
+	trackChatDisplayError: function() {
+		var { olark, isChatEnded } = this.state;
+
+		// If the user sent/received a message to/from an operator but the chatbox is not inlined
+		// then something must be going wrong. Lets record this event and give some details
```
Comment (dllh) [inline client/me/help/help-contact/index.jsx]: Minor typo. Should be `Let's`.

File: client/me/help/help-contact/index.jsx [hunk 2]
```diff
@@ -218,6 +222,33 @@ module.exports = React.createClass( {
...
+			olark_is_operator_available: olark.isOperatorAvailable
+		};
+
+		if ( olark.details ) {
+			tracksData.olark_is_conversing = olark.details.isConversing;
+		} else {
+			tracksData.olark_details_missing = true;
+		}
+
+		analytics.tracks.recordEvent( 'calypso_help_contact_chatbox_display_mistake', tracksData );
```
Comment (dllh) [inline client/me/help/help-contact/index.jsx]: I think the verb tends to go last in event names, so maybe something more like `calypso_help_contact_chatbox_mistaken_display`?

File: client/me/help/help-contact/index.jsx [hunk 3]
```diff
@@ -218,6 +222,33 @@ module.exports = React.createClass( {
...
+	trackChatDisplayError: function() {
+		var { olark, isChatEnded } = this.state;
+
+		// If the user sent/received a message to/from an operator but the chatbox is not inlined
+		// then something must be going wrong. Lets record this event and give some details
+		if ( this.canShowChatbox() ) {
+			return;
+		}
+
+		const tracksData = {
```
Comment (dllh) [inline client/me/help/help-contact/index.jsx]: Just for my education, why are you using a `const` here instead of a `var`?

**Comment by dllh** [comment]
One thing occurs to me: What if the error has to do with inconsistency of what we get back from `canShowChatbox()`? What if the root of the problem is actually with the bouncing value of `isOperatorAvailable`, for example? Will this solution overlook cases in which operator availability is the problem? 

I've wondered if maybe the issue didn't have to do with timing of availability toggling. So for example, maybe a user types a message at just about the same time that operators go away. Is it possible that this would get us in a state in which as operators become available and unavailable rapidly, we manage to keep changing state and somehow resending the messages?

Would it be more useful to just fire a Tracks ping for any relevant Olark action (message sent, availability change, command send, etc.) and then look at the history for users who trigger the bug and try to find patterns?

File: client/me/help/help-contact/index.jsx (hunk 3 above)
Comment (omarjackman) [inline client/me/help/help-contact/index.jsx]: The `var` on line 226 should've been `const` and I use it when the variable content never changes

**Comment by omarjackman** [comment]
@dllh

> One thing occurs to me: What if the error has to do with inconsistency of what we get back from canShowChatbox()?

My assumption is that the reason the chat is disappearing and the user is able to still see the contact form is because `canShowChatbox()` is returning `false`.

> What if the root of the problem is actually with the bouncing value of `isOperatorAvailable`?

This is why I'm adding all of that extra data in the tracks ping.

> Will this solution overlook cases in which operator availability is the problem?

No. It's aimed at catching the case where the contact form is still shown even though a chat is active regardless of operator availability.

File: client/me/help/help-contact/index.jsx (hunk 2 above)
Comment (dllh) [inline client/me/help/help-contact/index.jsx]: Any thoughts re the naming here? There's a fieldguide page about naming conventions that it'd be good to review and make sure we're following best practices.

**Comment by dllh** [comment]
The code here looks reasonable to me, though as you note it isn't really testable, unless you think we could do something with unit tests and firing the various events with the necessary state to trigger the Tracks ping. 

I've tried to think of ways in which this might introduce regressions or bugs, and I can't find any, so I think it's low-risk even if untestable. 

Please review the Tracks naming conventions to make sure we're doing what we should there. Otherwise, I think it'd be worth spending a few minutes seeing if unit tests would help, and if not, :ship:.

**Comment by omarjackman** [comment]
> Please review the Tracks naming conventions to make sure we're doing what we should there.

Changed it to `calypso_help_contact_chatbox_mistaken_display`

> I think it'd be worth spending a few minutes seeing if unit tests would help

It would be awesome to get some unit tests for this but it's not possible. Olarks api doesn't work under mocha and I currently have to mock the olark functions in order to have the tests that we do have.

**Comment by dllh** [comment]
Bummer about the tests. I'm not seeing any way in which this looks like it'd backfire, so even without being able to repro the functionality, I feel ok about offering a :ship: here.

**Pull request merged**